graphql_compiler/schema_generation/schema_graph.py

```diff
@@ -553,59 +544,74 @@ def _set_up_vertex_elements(self, class_name_to_definition):
+        }
 
         for property_definition in link_property_definitions:
-            property_name = property_definition['name']
-            links[property_name].append(self._create_descriptor_from_property_definition(
-                class_name, property_definition, class_name_to_definition))
-
-        for property_name in links:
-            elements = {
-                property_descriptor.qualifier
-                for property_descriptor in links[property_name]
-            }
-
-            # If there are multiple in/out properties, we choose to include the one that
-            # is a subclass of all the elements present in the in/out properties.
-            for property_descriptor in links[property_name]:
-                subclass_set = self._subclass_sets[property_descriptor.qualifier]
-                if len(elements.intersection(subclass_set)) == 1:
-                    current_descriptor = property_name_to_descriptor.get(property_name, None)
-                    if current_descriptor and current_descriptor != property_descriptor:
-                        raise AssertionError(u'There already exists property "{}" in addition '
-                                             u'to property "{}" which is a subclass of all '
-                                             u'in/out properties for class "{}".'
-                                             .format(current_descriptor,
-                                                     property_descriptor, class_name))
-                    property_name_to_descriptor[property_name] = property_descriptor
-
-            if (property_name not in property_name_to_descriptor and not abstract and
-                    issubclass(kind_cls, EdgeType)):
-                raise AssertionError(u'For property "{}" of non-abstract edge class "{}", '
+            self._validate_link_definition(class_name_to_definition, property_definition)
+            links[property_definition['name']].add(property_definition['linkedClass'])
+
+        for link_direction, linked_classes in six.iteritems(links):
+            for linked_class in linked_classes:
+                subclass_set = self._subclass_sets[linked_class]
+                if len(linked_classes.intersection(subclass_set)) == 1:
+                    base_direction_connection = base_connections.get(link_direction, None)
+                    if base_direction_connection and base_direction_connection != linked_class:
+                        raise AssertionError(u'There already exists class "{}" in addition '
+                                             u'to class "{}" which is a subclass of all '
+                                             u'{} properties for class "{}".'
+                                             .format(base_direction_connection,
+                                                     linked_class, link_direction, class_name))
+                    base_connections[link_direction] = linked_class
```

Comment:
Will refactor this in a future PR to make this function more understandable. I essentially performed some renaming here, but leaved the overall structure of the code intact.